File: typescript/tests/business_board.test.ts
// business_board.test.ts
// Verifies the CircleAI.Business port: unit hierarchy, latest-KPI lookup with
// NaN fallback, quarterly target achievement ratio, and context.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryBusinessBoard,
  BusinessDomainContext,
  businessUnit,
  kpiSample,
  quarterTarget,
} from "../src/business/index";

const D = (s: string) => new Date(s);

describe("InMemoryBusinessBoard", () => {
  it("adds units and finds children", () => {
    const b = new InMemoryBusinessBoard();
    b.add(businessUnit("root", "Group", "", ["rev"]));
    b.add(businessUnit("u1", "Sales", "root", ["rev"]));
    b.add(businessUnit("u2", "Eng", "root", ["velocity"]));
    assert.equal(b.getUnit("u1")?.name, "Sales");
    assert.equal(b.getUnit("nope"), undefined);
    assert.deepEqual(
      b.childrenOf("root").map((u) => u.unitId),
      ["u1", "u2"],
    );
  });

  it("returns the newest KPI value, or NaN when none", () => {
    const b = new InMemoryBusinessBoard();
    assert.ok(Number.isNaN(b.latestKpi("u1", "rev")));
    b.record(kpiSample("u1", "rev", 10, D("2026-01-01T00:00:00Z")));
    b.record(kpiSample("u1", "rev", 30, D("2026-03-01T00:00:00Z")));
    b.record(kpiSample("u1", "cost", 5, D("2026-03-01T00:00:00Z")));
    assert.equal(b.latestKpi("u1", "rev"), 30);
    assert.equal(b.latestKpi("u1", "cost"), 5);
  });

  it("computes target achievement, NaN when missing or zero target", () => {
    const b = new InMemoryBusinessBoard();
    b.record(kpiSample("u1", "rev", 75, D("2026-03-01T00:00:00Z")));
    b.setTarget(quarterTarget("u1", "rev", 2026, 1, 100));
    assert.equal(b.targetAchievement("u1", "rev", 2026, 1), 0.75);
    assert.ok(Number.isNaN(b.targetAchievement("u1", "rev", 2026, 2))); // no target
    b.setTarget(quarterTarget("u1", "rev", 2026, 3, 0));
    assert.ok(Number.isNaN(b.targetAchievement("u1", "rev", 2026, 3))); // zero target
  });

  it("rejects null arguments", () => {
    const b = new InMemoryBusinessBoard();
    assert.throws(() => b.add(null as never));
    assert.throws(() => b.record(null as never));
    assert.throws(() => b.setTarget(null as never));
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(BusinessDomainContext.systemPromptSnippet.includes("[DOMAIN: Business]"));
    assert.deepEqual(BusinessDomainContext.complianceFlags, ["POPIA", "Commercial_Law", "GDPR_aware"]);
    assert.deepEqual(BusinessDomainContext.suggestedTools, ["calendar", "web_search", "document_editor", "task_manager"]);
  });
});

File: go/tests/business_board_test.go
// business_board_test.go
//
// Verifies the CircleAI.Business port (business_board.go): unit add/get,
// ChildrenOf filter, KPI record + LatestKpi (most recent; NaN if none), target
// set + TargetAchievement (ratio; NaN when target missing or zero), and defensive
// KpiTags copy.

package circleai_test

import (
	"math"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestBusiness_UnitsAndChildren(t *testing.T) {
	b := circleai.NewInMemoryBusinessBoard()
	b.Add(circleai.BusinessUnit{UnitId: "root", Name: "Group", ParentUnitId: "", KpiTags: []string{"rev"}})
	b.Add(circleai.BusinessUnit{UnitId: "u1", Name: "Sales", ParentUnitId: "root"})
	b.Add(circleai.BusinessUnit{UnitId: "u2", Name: "Ops", ParentUnitId: "root"})

	if u, ok := b.GetUnit("u1"); !ok || u.Name != "Sales" {
		t.Fatalf("get u1 = %+v ok=%v", u, ok)
	}
	kids := b.ChildrenOf("root")
	if len(kids) != 2 || kids[0].UnitId != "u1" || kids[1].UnitId != "u2" {
		t.Fatalf("children order/count wrong: %+v", kids)
	}

	// Defensive KpiTags copy: mutating the caller's slice must not change stored.
	tags := []string{"rev"}
	b.Add(circleai.BusinessUnit{UnitId: "u3", Name: "Fin", ParentUnitId: "root", KpiTags: tags})
	tags[0] = "MUTATED"
	u3, _ := b.GetUnit("u3")
	if len(u3.KpiTags) != 1 || u3.KpiTags[0] != "rev" {
		t.Fatalf("KpiTags not defensively copied: %+v", u3.KpiTags)
	}
}

func TestBusiness_LatestKpiAndTarget(t *testing.T) {
	b := circleai.NewInMemoryBusinessBoard()
	t0 := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	b.Record(circleai.KpiSample{UnitId: "u1", Metric: "revenue", Value: 100, AtUtc: t0})
	b.Record(circleai.KpiSample{UnitId: "u1", Metric: "revenue", Value: 150, AtUtc: t0.Add(24 * time.Hour)})
	b.Record(circleai.KpiSample{UnitId: "u1", Metric: "cost", Value: 40, AtUtc: t0})

	if got := b.LatestKpi("u1", "revenue"); got != 150 {
		t.Fatalf("latest revenue = %v, want 150", got)
	}
	// No sample -> NaN.
	if got := b.LatestKpi("u1", "missing"); !math.IsNaN(got) {
		t.Fatalf("missing metric should be NaN, got %v", got)
	}

	b.SetTarget(circleai.QuarterTarget{UnitId: "u1", Metric: "revenue", Year: 2026, Quarter: 3, Target: 300})
	if got := b.TargetAchievement("u1", "revenue", 2026, 3); got != 0.5 { // 150/300
		t.Fatalf("achievement = %v, want 0.5", got)
	}
	// Missing target -> NaN.
	if got := b.TargetAchievement("u1", "revenue", 2026, 4); !math.IsNaN(got) {
		t.Fatalf("missing target should be NaN, got %v", got)
	}
	// Zero target -> NaN.
	b.SetTarget(circleai.QuarterTarget{UnitId: "u1", Metric: "cost", Year: 2026, Quarter: 3, Target: 0})
	if got := b.TargetAchievement("u1", "cost", 2026, 3); !math.IsNaN(got) {
		t.Fatalf("zero target should be NaN, got %v", got)
	}
}
